This notebook performs an Exploratory Data Analysis (EDA) on the Telco Customer Churn dataset

# Module 1 - Week 1
## Title: Exploratory Data Analysis: Telco Customer Churn
## Dataset: https://www.kaggle.com/datasets/blastchar/telco-customer-churn
## Course: ALY6040 80593 Data Mining Applications SEC 08 Spring 2025 CPS [TOR-A-HY]
### Group 1:
* Vijay Kumar Avula
* Divya Teja Kommineni
* Eugene Addo-Odei
* Mohammad Faizankhan Pathan
* Isaac Adjei

#### Date: April 23, 2025

In [2]:
import pandas as pd

In [3]:
# 1.1 Load and Inspect the Dataset
# Load the dataset
df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv', sep=',', encoding='utf-8')

# Output basic structure
print("Dataset Shape:", df.shape)
print("\nColumns in the Dataset:\n", df.columns.tolist())
print("\nFirst 5 Rows:\n", df.head())
print("\nLast 5 Rows:\n", df.tail())

Dataset Shape: (7043, 21)

Columns in the Dataset:
 ['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn']

First 5 Rows:
    customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No 

In [ ]:
**Immediate Observations:**


1.   The dataset contains **7043 rows** and **21 columns**.
2.   Features include both categorical (e.g., *gender, Partner, Churn*) and numerical variables (e.g., *tenure, MonthlyCharges*).
3.   Potential issues:

  *   *TotalCharges* appears to be numeric but might contain non-numeric values (e.g., empty strings).
  *   *SeniorCitizen* is binary (0/1), which could be treated as categorical.
  *   *customerID* is an identifier column and may not contribute to analysis.

In [12]:
# 1.2 Feature Catalog & Data Dictionary
# Function to generate the feature catalog
def generate_feature_catalog(df):
    # Initialize an empty list to store feature details
    feature_catalog = []

    # Loop through each column in the DataFrame
    for col in df.columns:
        # Determine the data type of the column
        dtype = df[col].dtype

        # Count unique values for categorical columns
        if dtype == 'object':
            unique_values = df[col].nunique()
        else:
            unique_values = None  # Not applicable for numerical columns

        # Add a description based on the column name (customizable)
        description = ""
        if col == "customerID":
            description = "Unique identifier for each customer"
        elif col == "gender":
            description = "Customer’s gender (Male/Female)"
        elif col == "SeniorCitizen":
            description = "Whether the customer is a senior citizen (0/1)"
        elif col == "Partner":
            description = "Whether the customer has a partner (Yes/No)"
        elif col == "Dependents":
            description = "Whether the customer has dependents (Yes/No)"
        elif col == "tenure":
            description = "Number of months the customer has been with the company"
        elif col == "PhoneService":
            description = "Whether the customer has phone service (Yes/No)"
        elif col == "MultipleLines":
            description = "Whether the customer has multiple phone lines (Yes/No/No phone service)"
        elif col == "InternetService":
            description = "Type of internet service (DSL/Fiber optic/No)"
        elif col == "OnlineSecurity":
            description = "Whether the customer has online security (Yes/No/No internet service)"
        elif col == "OnlineBackup":
            description = "Whether the customer has online backup (Yes/No/No internet service)"
        elif col == "DeviceProtection":
            description = "Whether the customer has device protection (Yes/No/No internet service)"
        elif col == "TechSupport":
            description = "Whether the customer has tech support (Yes/No/No internet service)"
        elif col == "StreamingTV":
            description = "Whether the customer has streaming TV (Yes/No/No internet service)"
        elif col == "StreamingMovies":
            description = "Whether the customer has streaming movies (Yes/No/No internet service)"
        elif col == "Contract":
            description = "Type of contract (Month-to-month/One year/Two year)"
        elif col == "PaperlessBilling":
            description = "Whether the customer uses paperless billing (Yes/No)"
        elif col == "PaymentMethod":
            description = "Payment method (Electronic check/Mailed check/Bank transfer/Credit card)"
        elif col == "MonthlyCharges":
            description = "Monthly charges for services"
        elif col == "TotalCharges":
            description = "Total charges for services over the tenure"
        elif col == "Churn":
            description = "Whether the customer churned (Yes/No)"

        # Append feature details to the catalog
        feature_catalog.append({
            "Feature Name": col,
            "Data Type": dtype,
            "Unique Values": unique_values,
            "Description": description
        })

    # Convert the catalog to a DataFrame for better visualization
    feature_catalog_df = pd.DataFrame(feature_catalog)
    return feature_catalog_df

# Generate the feature catalog
feature_catalog_df = generate_feature_catalog(df)

# Display the feature catalog
print(feature_catalog_df)

        Feature Name Data Type  Unique Values  \
0         customerID    object         7043.0   
1             gender    object            2.0   
2      SeniorCitizen     int64            NaN   
3            Partner    object            2.0   
4         Dependents    object            2.0   
5             tenure     int64            NaN   
6       PhoneService    object            2.0   
7      MultipleLines    object            3.0   
8    InternetService    object            3.0   
9     OnlineSecurity    object            3.0   
10      OnlineBackup    object            3.0   
11  DeviceProtection    object            3.0   
12       TechSupport    object            3.0   
13       StreamingTV    object            3.0   
14   StreamingMovies    object            3.0   
15          Contract    object            3.0   
16  PaperlessBilling    object            2.0   
17     PaymentMethod    object            4.0   
18    MonthlyCharges   float64            NaN   
19      TotalCharges

**Key Observations:**
  * **Numerical Features:**
    * SeniorCitizen, tenure, MonthlyCharges.
  * Categorical Features:
    * Most other features are categorical (object type).
    * Special Case: TotalCharges is stored as object due to empty strings (" ").
  * Unique Values:
    * Binary Features: gender, Partner, Dependents, PhoneService, PaperlessBilling, Churn.
    * Multi-Class Categorical Features:
      * MultipleLines, InternetService, OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV, StreamingMovies, Contract: 3 unique values each.
      * PaymentMethod: 4 unique values.

In [13]:
# 1.3 Initial Value Checks and Missing Data
# Check for missing data
missing_data = df.isnull().sum()
missing_percentage = (missing_data / len(df)) * 100

# Unique values in categorical columns
unique_values = {col: df[col].nunique() for col in df.select_dtypes(include=['object']).columns}

# Identify potential issues in TotalCharges
total_charges_issues = df[df['TotalCharges'] == " "]

print("Missing Data:\n", missing_data)
print("\nMissing Percentage:\n", missing_percentage)
print("\nUnique Values:\n", unique_values)
print("\nIssues in TotalCharges:\n", total_charges_issues)

Missing Data:
 customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

Missing Percentage:
 customerID          0.0
gender              0.0
SeniorCitizen       0.0
Partner             0.0
Dependents          0.0
tenure              0.0
PhoneService        0.0
MultipleLines       0.0
InternetService     0.0
OnlineSecurity      0.0
OnlineBackup        0.0
DeviceProtection    0.0
TechSupport         0.0
StreamingTV         0.0
StreamingMovies     0.0
Contract            0.0
PaperlessBilling    0.0
PaymentMethod       0.0
MonthlyCharges      0.0
TotalCharges        0.0
Churn   

**Missing Data Analysis:**
  * .isnull().sum() shows no explicit NaN values in any column.
  * However, TotalCharges contains empty strings (" ") in 11 rows , which are not recognized as missing values by pandas.
  *These rows correspond to customers with tenure = 0, indicating they are new customers who have not yet incurred charges.

**Unique Values:**
  * Confirmed logical unique value counts for all categorical features.
  * Numerical features (tenure, MonthlyCharges, TotalCharges) do not have fixed unique values due to their continuous nature.

**Data Integrity:**
  * No other suspicious or illogical values were identified.

#### 1.4 Early Questions & Hypothesis Generation
##### **Hypotheses:**
1.   Higher Monthly Charges Lead to Churn: Customers paying higher monthly fees are more likely to churn due to affordability concerns.
  * Rationale: Higher costs may exceed perceived value.
2.   Tenure is Inversely Related to Churn: Long-term customers are less likely to churn.
  * Rationale: Loyal customers tend to stay longer.
3.  Fiber Optic Internet Increases Churn: Customers with fiber optic internet may churn more due to higher costs or technical issues.
  * Rationale: Fiber optic plans are often pricier and may have reliability concerns.
4.  Lack of Tech Support Leads to Churn: Customers without tech support are more likely to churn due to unresolved issues.
  * Rationale: Tech support improves customer satisfaction.
5.  Paperless Billing Affects Churn: Customers using paperless billing may churn more due to billing transparency issues.
  * Rationale: Some customers prefer traditional billing methods.
6.  Senior Citizens Churn More: Older customers may churn more due to difficulty adapting to new technologies.
  * Rationale: Age-related challenges may impact retention.
7.  Short-Term Contracts Increase Churn: Month-to-month contracts lead to higher churn rates compared to long-term contracts.
  * Rationale: Short-term contracts lack commitment.

In [5]:
# Overview of missing values
print("Missing values in each column:\n", df.isnull().sum())

# CATEGORICAL IMPUTATION (Mode)
ategorical_cols = ['Partner', 'Dependents', 'MultipleLines']
for col in categorical_cols:
    mode_val = df[col].mode()[0]
    df[col] = df[col].fillna(mode_val) 
    print(f"Imputed '{col}' with mode: {mode_val}")

# NUMERICAL IMPUTATION (Median or Mean for TotalCharges)
# First, check the data type and convert if necessary
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
median_total = df['TotalCharges'].median()
df['TotalCharges'] = df['TotalCharges'].fillna(median_total) 
print(f"Imputed 'TotalCharges' with median: {median_total}")

# Verify results
print("Remaining missing values:\n", df.isnull().sum())



Missing values in each column:
 customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64
Imputed 'Partner' with mode: No
Imputed 'Dependents' with mode: No
Imputed 'MultipleLines' with mode: No
Imputed 'TotalCharges' with median: 1397.475
Remaining missing values:
 customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
Streamin

Feature-Specific Strategy:

- **Initial Missing Values Detected**:
  - `Partner`, `Dependents`, `MultipleLines`: Categorical
  - `TotalCharges`: Numeric (due to blank entries or conversion issues)

- **Imputation Strategy**:
  - `Partner`, `Dependents`, `MultipleLines`: Imputed using **mode** (`'No'`)
  - `TotalCharges`: Imputed using **median** (**1397.475**)

- **Final Status**:
  - **All missing values successfully handled**
  - No missing data remains in any column

The dataset is now complete and ready for analysis or modelling.



In [7]:
# Check for negative values
neg_total = df[df['TotalCharges'] < 0]
neg_monthly = df[df['MonthlyCharges'] < 0]

print(f"Negative values in TotalCharges: {len(neg_total)}")
print(f"Negative values in MonthlyCharges: {len(neg_monthly)}")

# Only proceed if negative values exist
if len(neg_total) + len(neg_monthly) > 0:
    # Calculate medians from non-negative values
    median_total = df.loc[df['TotalCharges'] >= 0, 'TotalCharges'].median()
    median_monthly = df.loc[df['MonthlyCharges'] >= 0, 'MonthlyCharges'].median()

    # Replace negatives
    df.loc[df['TotalCharges'] < 0, 'TotalCharges'] = median_total
    df.loc[df['MonthlyCharges'] < 0, 'MonthlyCharges'] = median_monthly
    
    print("Negative values replaced with medians.")
else:
    print("No negative values found - no replacement needed.")

# Verify results
print("\nPost-cleaning summary:")
print(f"TotalCharges min: {df['TotalCharges'].min():.2f}")
print(f"MonthlyCharges min: {df['MonthlyCharges'].min():.2f}")

Negative values in TotalCharges: 0
Negative values in MonthlyCharges: 0
No negative values found - no replacement needed.

Post-cleaning summary:
TotalCharges min: 18.80
MonthlyCharges min: 18.25


Report on Handling Erroneous Values:

- **Negative Value Check**:
  - `TotalCharges`: **0** negative values
  - `MonthlyCharges`: **0** negative values
- **No replacements needed** — all values are logically valid.

**Post-cleaning summary:**
- `TotalCharges` minimum: 18.80
- `MonthlyCharges` minimum: 18.25

Billing data confirmed clean and consistent.


In [9]:
from sklearn.preprocessing import LabelEncoder

# 1. Convert 'Churn' to binary (clear explicit mapping)
df['Churn'] = df['Churn'].map({'No': 0, 'Yes': 1})

# 2. Ensure numeric columns are properly typed (add float conversion)
numeric_cols = ['tenure', 'MonthlyCharges', 'TotalCharges']
df[numeric_cols] = df[numeric_cols].apply(pd.to_numeric, errors='coerce')

# 3. Identify categorical features (include category dtype and check cardinality)
cat_features = df.select_dtypes(include=['object', 'category']).columns.tolist()
print("Initial categorical features:", cat_features)

# 4. Handle constant columns (nunique == 1) before encoding
constant_cols = [col for col in cat_features if df[col].nunique() == 1]
if constant_cols:
    print(f"Dropping constant columns: {constant_cols}")
    df.drop(columns=constant_cols, inplace=True)
    cat_features = [col for col in cat_features if col not in constant_cols]

# 5. Split remaining categorical features
binary_cols = [col for col in cat_features if df[col].nunique() == 2]
multi_cat_cols = [col for col in cat_features if df[col].nunique() > 2]

# 6. Safer binary encoding using factorize instead of LabelEncoder
for col in binary_cols:
    df[col], _ = pd.factorize(df[col])
    print(f"Binary encoded: {col} (0={df[col].unique()[0]}, 1={df[col].unique()[1]})")

# 7. One-Hot Encoding with prefix for clarity
if multi_cat_cols:
    df = pd.get_dummies(
        df, 
        columns=multi_cat_cols,
        prefix=multi_cat_cols,
        drop_first=True,
        dtype='uint8'  # Save memory
    )
    print(f"One-hot encoded: {multi_cat_cols}")

# 8. Final validation
remaining_cats = df.select_dtypes(include=['object', 'category']).columns
print("\nFinal check - remaining categorical features:", remaining_cats.tolist())
print("Data shape after encoding:", df.shape)

Initial categorical features: ['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod']
Binary encoded: gender (0=0, 1=1)
Binary encoded: Partner (0=0, 1=1)
Binary encoded: Dependents (0=0, 1=1)
Binary encoded: PhoneService (0=0, 1=1)
Binary encoded: PaperlessBilling (0=0, 1=1)
One-hot encoded: ['customerID', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaymentMethod']

Final check - remaining categorical features: []
Data shape after encoding: (7043, 7073)


Erroneous Values Handling Report

1. TotalCharges & MonthlyCharges: Converted to numeric.

2. Negative values: Replaced with a median of valid entries.

3. Missing or invalid TotalCharges: Also imputed with median.

4. Result: Clean, non-negative billing data.


Categorical Encoding Summary

1. Label Encoded (binary): gender, Partner, Dependents, PhoneService, PaperlessBilling

2. One-Hot Encoded: MultipleLines, InternetService, OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV, StreamingMovies, Contract, PaymentMethod, customerID

3. All categorical features are encoded.

4. Final dataset shape: (7043, 7073).